### ISSUE TITLE
xkb bug

### ISSUE THREAD

**Issue by MarkLTZ** [TigerVNC/tigervnc#1000]
Hello,
I found that starting tigervnc-server do not load xkb settings from the system defaults but it always start in US.

```
[root@localhost tmp]# setxkbmap -print -verbose 10
Setting verbose level to 10
locale is C
Trying to load rules file ./rules/evdev...
Trying to load rules file /usr/share/X11/xkb/rules/evdev...
Success.
Applied rules from evdev:
rules:      evdev
model:      pc105
layout:     us
Trying to build keymap using the following components:
keycodes:   evdev+aliases(qwerty)
types:      complete
compat:     complete
symbols:    pc+us+inet(evdev)
geometry:   pc(pc105)
xkb_keymap {
	xkb_keycodes  { include "evdev+aliases(qwerty)"	};
	xkb_types     { include "complete"	};
	xkb_compat    { include "complete"	};
	xkb_symbols   { include "pc+us+inet(evdev)"	};
	xkb_geometry  { include "pc(pc105)"	};
};
```
I set my keyboard language with the command `setxkbmap -layout it` but the change is not persistent, so if i just restart my vncviewer session, the xkb layout is resetted to US.

This make apps using xkb for example freerdp not working for non US keyboard.

Could you please help me? How to setup xkb persistence for another keyboard language? If not, while waiting for a new feature, there is an hack in the code for using a fixed value different from US (example IT) ?

Thanks,
Mark

**Comment by CendioOssman** [comment]
I'm afraid `setxkbmap` does not store anything and any changes it makes are only temporary.

Setting the correct layout is generally handled by your desktop environment so please have a look there. Other than that you'll have to put that `setxkbmap` call in your session startup script and it should set up things correctly.

**Comment by MarkLTZ** [comment]
My desktop environment is configured using Italian keyboard and as I said before program affected are only the one based on xkb scanning. I repeat vncserver comes up always with US layout keyboard or is reset to US at every client connection.

Please **re-open this issue** and **investigate at least few minutes** on my issue before closing it.

If you need more details, please ask. Closing the issue without taking care of my problem is very bad.

Thanks

**Comment by MarkLTZ** [comment]
I forgot.... issue happen on RHEL6

**Comment by MarkLTZ** [comment]
here the evidence:

```
[mark@localhost ~]$ export DISPLAY=:0
[mark@localhost ~]$ setxkbmap -print -verbose 10
Setting verbose level to 10
locale is C
Trying to load rules file ./rules/evdev...
Trying to load rules file /usr/share/X11/xkb/rules/evdev...
Success.
Applied rules from evdev:
rules:      evdev
model:      pc105
**layout:     it**
options:    terminate:ctrl_alt_bksp
Trying to build keymap using the following components:
keycodes:   evdev+aliases(qwerty)
types:      complete
compat:     complete
symbols:    pc+it+inet(evdev)+terminate(ctrl_alt_bksp)
geometry:   pc(pc105)
xkb_keymap {
	xkb_keycodes  { include "evdev+aliases(qwerty)"	};
	xkb_types     { include "complete"	};
	xkb_compat    { include "complete"	};
	xkb_symbols   { include "pc+it+inet(evdev)+terminate(ctrl_alt_bksp)"	};
	xkb_geometry  { include "pc(pc105)"	};
};

[mark@localhost ~]$ vncserver -depth 16 -geometry 1900x1600 -AlwaysShared -SecurityTypes VncAuth 

New 'localhost.localdomain:1 (mark)' desktop is localhost.localdomain:1

Creating default startup script /home/mark/.vnc/xstartup
Starting applications specified in /home/mark/.vnc/xstartup
Log file is /home/mark/.vnc/localhost.localdomain:1.log

[mark@localhost ~]$ export DISPLAY=:1

[mark@localhost ~]$ setxkbmap -print -verbose 10
Setting verbose level to 10
locale is C
Trying to load rules file ./rules/evdev...
Trying to load rules file /usr/share/X11/xkb/rules/evdev...
Success.
Applied rules from evdev:
rules:      evdev
model:      pc105
**layout:     us**
Trying to build keymap using the following components:
keycodes:   evdev+aliases(qwerty)
types:      complete
compat:     complete
symbols:    pc+us+inet(evdev)
geometry:   pc(pc105)
xkb_keymap {
	xkb_keycodes  { include "evdev+aliases(qwerty)"	};
	xkb_types     { include "complete"	};
	xkb_compat    { include "complete"	};
	xkb_symbols   { include "pc+us+inet(evdev)"	};
	xkb_geometry  { include "pc(pc105)"	};
};
```

**Comment by CendioOssman** [comment]
I'm afraid the issue is somewhat well known and unfortunately not under our control. So there isn't anything for us in TigerVNC to investigate. I would expect the desktop environment to change the layout to Italian when it starts up in your VNC session. Since that doesn't happen I would normally ask you to report the issue to them. However RHEL 6 is pretty much EOL so I doubt Red Hat will address this.

Instead you should be able to work around the issue by placing a call to `setxkbmap` in your startup script. I.e. `/home/mark/.vnc/xstartup`.

**Comment by MarkLTZ** [comment]
Placing setxkbmap inside xstartup do not solve the issue because a reset to US occurs at every client connection.

For the moment I found a workaround starting gnome inside the xstartup, but it could be nice to have a feature to set the keyboard layout at runtime (client or server) because in our case we do not need to start the gnome but only the final tools like freerdp and others to have them "kiosked" inside the X11 of tigervnc. Having also gnome running is a potential breach in the security of our application that offers to our customers remote access to systems/tools from web using this flow: browser --> webapp (rfb) --> vncserver --> tool (freerdp / others)

Of course the issue above was discovered by testing only the flow vncclient -> vcnserver -> freerdp

Because at the moment as I said before I found a workaround, I would like to ask to you to think to develop a new feature of selecting default xkb lauout at startup of vncserver and/or as parameter for the vncclient.

Thanks.
Mark
